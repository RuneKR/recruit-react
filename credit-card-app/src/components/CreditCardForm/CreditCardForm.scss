.creditCardForm-form {
  max-width: 320px;
  margin: auto;
  padding: 10px;

  .ant-form-item-label {
    > label::after {
      display: none;
    }
  }
}

.creditCardForm-cardNumberFormItem {
  margin-bottom: 0;
}

.creditCardForm-cvcExpiryRow {
  display: flex;
  justify-content: space-between;

  > .ant-form-item {
    &:first-child {
      margin-right: 3px;
    }
    &:nth-child(2) {
      margin-left: 3px;
    }
  }
}

.creditCardForm-submitButton {
  width: 100%;
}
