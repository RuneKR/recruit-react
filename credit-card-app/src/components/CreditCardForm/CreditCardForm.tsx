import React, { Component } from "react";
import { Input, Button } from "antd/lib";
import Form, { FormComponentProps } from "antd/lib/form";

import "./CreditCardForm.scss";
import {
  formatCardExpiryDate,
  formatCardNumber,
  formatCvc
} from "../../common/formatters";
import { CreditCard } from "../../models/creditCard.model";

type State = typeof initialState;
type OwnProps = {
  onSubmit: (creditCard: CreditCard) => void;
};
type Props = FormComponentProps & OwnProps;

const initialState = {
  submitClicked: false
};

class CreditCardForm extends Component<Props, State> {
  constructor(props: Props) {
    super(props);
    this.state = initialState;
  }

  formatCvcField = (e: React.ChangeEvent<HTMLInputElement>) => {
    return formatCvc(e.target.value);
  };

  formatCardNumberField = (e: React.ChangeEvent<HTMLInputElement>) => {
    return formatCardNumber(e.target.value);
  };

  formatExpiryDateField = (e: React.ChangeEvent<HTMLInputElement>) => {
    let value = e.target.value;
    if ((e.nativeEvent as any).inputType != "deleteContentBackward") {
      value = formatCardExpiryDate(value);
    }
    return value;
  };

  handleSubmit = () => {
    this.setState({ submitClicked: true });
    this.props.form.validateFields((err: any, values: any) => {
      if (!err) {
        this.props.onSubmit(values);
      }
    });
  };

  render() {
    const { getFieldDecorator } = this.props.form;

    return (
      <Form className="creditCardForm-form">
        <Form.Item
          label="Card Number"
          className="creditCardForm-cardNumberFormItem"
        >
          {getFieldDecorator("number", {
            getValueFromEvent: this.formatCardNumberField,
            rules: [
              { required: true, message: "Required" },
              { len: 19, message: "Too short" }
            ],
            validateTrigger: this.state.submitClicked ? "onChange" : "onSubmit"
          })(<Input placeholder="1234 1234 1234 1234" maxLength={19} />)}
        </Form.Item>
        <div className="creditCardForm-cvcExpiryRow">
          <Form.Item label="CVC" className="creditCardForm-cvcFormItem">
            {getFieldDecorator("cvc", {
              getValueFromEvent: this.formatCvcField,
              rules: [
                { required: true, message: "Required" },
                { len: 3, message: "Too short" }
              ],
              validateTrigger: this.state.submitClicked
                ? "onChange"
                : "onSubmit"
            })(<Input placeholder="123" maxLength={3} />)}
          </Form.Item>
          <Form.Item label="Expiry" className="creditCardForm-expiryFormItem">
            {getFieldDecorator("expiry", {
              getValueFromEvent: this.formatExpiryDateField,
              rules: [
                { required: true, message: "Required" },
                { len: 5, message: "Too short" }
              ],
              validateTrigger: this.state.submitClicked
                ? "onChange"
                : "onSubmit"
            })(<Input placeholder="12/34" maxLength={5} />)}
          </Form.Item>
        </div>
        <Form.Item>
          <Button
            className="creditCardForm-submitButton"
            onClick={this.handleSubmit}
          >
            Submit
          </Button>
        </Form.Item>
      </Form>
    );
  }
}

export default Form.create()(CreditCardForm);
