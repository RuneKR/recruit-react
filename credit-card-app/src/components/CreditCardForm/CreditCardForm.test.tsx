import React from "react";
import renderer from "react-test-renderer";
import Enzyme, { shallow, mount } from "enzyme";
import Adapter from "enzyme-adapter-react-16";
import CreditCardForm from "./CreditCardForm";

Enzyme.configure({ adapter: new Adapter() });

describe("CreditCardForm Component", () => {
  test("form renders correctly", () => {
    const component = renderer.create(<CreditCardForm onSubmit={() => {}} />);
    let tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  test("required messages appear", () => {
    const component = mount(<CreditCardForm onSubmit={() => {}} />);

    component.find("Button").simulate("click");

    expect(
      component
        .find(".creditCardForm-cardNumberFormItem .has-error .ant-form-explain")
        .text()
    ).toEqual("Required");

    expect(
      component
        .find(".creditCardForm-cvcFormItem .has-error .ant-form-explain")
        .text()
    ).toEqual("Required");

    expect(
      component
        .find(".creditCardForm-expiryFormItem .has-error .ant-form-explain")
        .text()
    ).toEqual("Required");
  });

  test("too short messages appear", () => {
    const component = mount(<CreditCardForm onSubmit={() => {}} />);

    component
      .find(".creditCardForm-cardNumberFormItem input")
      .simulate("change", { target: { value: "0" } });

    component
      .find(".creditCardForm-cvcFormItem input")
      .simulate("change", { target: { value: "0" } });

    component
      .find(".creditCardForm-expiryFormItem input")
      .simulate("change", { target: { value: "0" } });

    component.find("Button").simulate("click");

    expect(
      component
        .find(".creditCardForm-cardNumberFormItem .has-error .ant-form-explain")
        .text()
    ).toEqual("Too short");

    expect(
      component
        .find(".creditCardForm-cvcFormItem .has-error .ant-form-explain")
        .text()
    ).toEqual("Too short");

    expect(
      component
        .find(".creditCardForm-expiryFormItem .has-error .ant-form-explain")
        .text()
    ).toEqual("Too short");
  });

  test("card number field formatting", () => {
    const component = mount(<CreditCardForm onSubmit={() => {}} />);
    const selector = ".creditCardForm-cardNumberFormItem input";

    component
      .find(selector)
      .simulate("change", { target: { value: "1234123412341234" } });
    expect(component.find(selector).props().value).toEqual(
      "1234 1234 1234 1234"
    );

    component
      .find(selector)
      .simulate("change", { target: { value: "fdasklfj" } });
    expect(component.find(selector).props().value).toEqual("");

    component
      .find(selector)
      .simulate("change", { target: { value: "12xxx341234#123&&." } });
    expect(component.find(selector).props().value).toEqual("1234 1234 123");
  });

  test("cvc field formatting", () => {
    const component = mount(<CreditCardForm onSubmit={() => {}} />);
    const selector = ".creditCardForm-cvcFormItem input";

    component
      .find(selector)
      .simulate("change", { target: { value: "1#@!  2 ##...3" } });
    expect(component.find(selector).props().value).toEqual("123");

    component.find(selector).simulate("change", { target: { value: "dasds" } });
    expect(component.find(selector).props().value).toEqual("");
  });

  test("expiry field formatting", () => {
    const component = mount(<CreditCardForm onSubmit={() => {}} />);
    const selector = ".creditCardForm-expiryFormItem input";

    component.find(selector).simulate("change", { target: { value: "3/" } });
    expect(component.find(selector).props().value).toEqual("03/");

    component.find(selector).simulate("change", { target: { value: "11" } });
    expect(component.find(selector).props().value).toEqual("11/");

    component.find(selector).simulate("change", { target: { value: "13" } });
    expect(component.find(selector).props().value).toEqual("01/3");

    component.find(selector).simulate("change", { target: { value: "1/11" } });
    expect(component.find(selector).props().value).toEqual("01/11");

    component.find(selector).simulate("change", { target: { value: "0/" } });
    expect(component.find(selector).props().value).toEqual("0");

    component.find(selector).simulate("change", { target: { value: "00" } });
    expect(component.find(selector).props().value).toEqual("0");

    component.find(selector).simulate("change", { target: { value: "a" } });
    expect(component.find(selector).props().value).toEqual("");

    component.find(selector).simulate("change", { target: { value: "01//" } });
    expect(component.find(selector).props().value).toEqual("01/");
  });
});
