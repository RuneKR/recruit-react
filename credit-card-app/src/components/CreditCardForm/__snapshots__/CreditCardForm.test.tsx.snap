// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CreditCardForm Component form renders correctly 1`] = `
<form
  className="ant-form ant-form-horizontal creditCardForm-form"
  onSubmit={[Function]}
>
  <div
    className="ant-row ant-form-item creditCardForm-cardNumberFormItem"
  >
    <div
      className="ant-col ant-form-item-label"
    >
      <label
        className="ant-form-item-required"
        htmlFor="number"
        onClick={[Function]}
        title="Card Number"
      >
        Card Number
      </label>
    </div>
    <div
      className="ant-col ant-form-item-control-wrapper"
    >
      <div
        className="ant-form-item-control"
      >
        <span
          className="ant-form-item-children"
        >
          <input
            className="ant-input"
            data-__field={
              Object {
                "name": "number",
              }
            }
            data-__meta={
              Object {
                "getValueFromEvent": [Function],
                "name": "number",
                "originalProps": Object {
                  "disabled": false,
                  "maxLength": 19,
                  "placeholder": "1234 1234 1234 1234",
                  "type": "text",
                },
                "ref": null,
                "rules": Array [
                  Object {
                    "message": "Required",
                    "required": true,
                  },
                  Object {
                    "len": 19,
                    "message": "Too short",
                  },
                ],
                "trigger": "onChange",
                "validate": Array [
                  Object {
                    "rules": Array [
                      Object {
                        "message": "Required",
                        "required": true,
                      },
                      Object {
                        "len": 19,
                        "message": "Too short",
                      },
                    ],
                    "trigger": Array [
                      "onSubmit",
                    ],
                  },
                ],
                "validateTrigger": "onSubmit",
                "valuePropName": "value",
              }
            }
            disabled={false}
            id="number"
            maxLength={19}
            onChange={[Function]}
            onKeyDown={[Function]}
            onSubmit={[Function]}
            placeholder="1234 1234 1234 1234"
            type="text"
            value=""
          />
        </span>
      </div>
    </div>
  </div>
  <div
    className="creditCardForm-cvcExpiryRow"
  >
    <div
      className="ant-row ant-form-item creditCardForm-cvcFormItem"
    >
      <div
        className="ant-col ant-form-item-label"
      >
        <label
          className="ant-form-item-required"
          htmlFor="cvc"
          onClick={[Function]}
          title="CVC"
        >
          CVC
        </label>
      </div>
      <div
        className="ant-col ant-form-item-control-wrapper"
      >
        <div
          className="ant-form-item-control"
        >
          <span
            className="ant-form-item-children"
          >
            <input
              className="ant-input"
              data-__field={
                Object {
                  "name": "cvc",
                }
              }
              data-__meta={
                Object {
                  "getValueFromEvent": [Function],
                  "name": "cvc",
                  "originalProps": Object {
                    "disabled": false,
                    "maxLength": 3,
                    "placeholder": "123",
                    "type": "text",
                  },
                  "ref": null,
                  "rules": Array [
                    Object {
                      "message": "Required",
                      "required": true,
                    },
                    Object {
                      "len": 3,
                      "message": "Too short",
                    },
                  ],
                  "trigger": "onChange",
                  "validate": Array [
                    Object {
                      "rules": Array [
                        Object {
                          "message": "Required",
                          "required": true,
                        },
                        Object {
                          "len": 3,
                          "message": "Too short",
                        },
                      ],
                      "trigger": Array [
                        "onSubmit",
                      ],
                    },
                  ],
                  "validateTrigger": "onSubmit",
                  "valuePropName": "value",
                }
              }
              disabled={false}
              id="cvc"
              maxLength={3}
              onChange={[Function]}
              onKeyDown={[Function]}
              onSubmit={[Function]}
              placeholder="123"
              type="text"
              value=""
            />
          </span>
        </div>
      </div>
    </div>
    <div
      className="ant-row ant-form-item creditCardForm-expiryFormItem"
    >
      <div
        className="ant-col ant-form-item-label"
      >
        <label
          className="ant-form-item-required"
          htmlFor="expiry"
          onClick={[Function]}
          title="Expiry"
        >
          Expiry
        </label>
      </div>
      <div
        className="ant-col ant-form-item-control-wrapper"
      >
        <div
          className="ant-form-item-control"
        >
          <span
            className="ant-form-item-children"
          >
            <input
              className="ant-input"
              data-__field={
                Object {
                  "name": "expiry",
                }
              }
              data-__meta={
                Object {
                  "getValueFromEvent": [Function],
                  "name": "expiry",
                  "originalProps": Object {
                    "disabled": false,
                    "maxLength": 5,
                    "placeholder": "12/34",
                    "type": "text",
                  },
                  "ref": null,
                  "rules": Array [
                    Object {
                      "message": "Required",
                      "required": true,
                    },
                    Object {
                      "len": 5,
                      "message": "Too short",
                    },
                  ],
                  "trigger": "onChange",
                  "validate": Array [
                    Object {
                      "rules": Array [
                        Object {
                          "message": "Required",
                          "required": true,
                        },
                        Object {
                          "len": 5,
                          "message": "Too short",
                        },
                      ],
                      "trigger": Array [
                        "onSubmit",
                      ],
                    },
                  ],
                  "validateTrigger": "onSubmit",
                  "valuePropName": "value",
                }
              }
              disabled={false}
              id="expiry"
              maxLength={5}
              onChange={[Function]}
              onKeyDown={[Function]}
              onSubmit={[Function]}
              placeholder="12/34"
              type="text"
              value=""
            />
          </span>
        </div>
      </div>
    </div>
  </div>
  <div
    className="ant-row ant-form-item"
  >
    <div
      className="ant-col ant-form-item-control-wrapper"
    >
      <div
        className="ant-form-item-control"
      >
        <span
          className="ant-form-item-children"
        >
          <button
            className="ant-btn creditCardForm-submitButton"
            onClick={[Function]}
            type="button"
          >
            <span>
              Submit
            </span>
          </button>
        </span>
      </div>
    </div>
  </div>
</form>
`;

exports[`Form renders correctly 1`] = `
<form
  className="ant-form ant-form-horizontal creditCardForm-form"
  onSubmit={[Function]}
>
  <div
    className="ant-row ant-form-item creditCardForm-cardNumberFormItem"
  >
    <div
      className="ant-col ant-form-item-label"
    >
      <label
        className="ant-form-item-required"
        htmlFor="number"
        onClick={[Function]}
        title="Card Number"
      >
        Card Number
      </label>
    </div>
    <div
      className="ant-col ant-form-item-control-wrapper"
    >
      <div
        className="ant-form-item-control"
      >
        <span
          className="ant-form-item-children"
        >
          <input
            className="ant-input"
            data-__field={
              Object {
                "name": "number",
              }
            }
            data-__meta={
              Object {
                "getValueFromEvent": [Function],
                "name": "number",
                "originalProps": Object {
                  "disabled": false,
                  "maxLength": 19,
                  "placeholder": "1234 1234 1234 1234",
                  "type": "text",
                },
                "ref": null,
                "rules": Array [
                  Object {
                    "message": "Required",
                    "required": true,
                  },
                  Object {
                    "len": 19,
                    "message": "Too short",
                  },
                ],
                "trigger": "onChange",
                "validate": Array [
                  Object {
                    "rules": Array [
                      Object {
                        "message": "Required",
                        "required": true,
                      },
                      Object {
                        "len": 19,
                        "message": "Too short",
                      },
                    ],
                    "trigger": Array [
                      "onSubmit",
                    ],
                  },
                ],
                "validateTrigger": "onSubmit",
                "valuePropName": "value",
              }
            }
            disabled={false}
            id="number"
            maxLength={19}
            onChange={[Function]}
            onKeyDown={[Function]}
            onSubmit={[Function]}
            placeholder="1234 1234 1234 1234"
            type="text"
            value=""
          />
        </span>
      </div>
    </div>
  </div>
  <div
    className="creditCardForm-cvcExpiryRow"
  >
    <div
      className="ant-row ant-form-item creditCardForm-cvcFormItem"
    >
      <div
        className="ant-col ant-form-item-label"
      >
        <label
          className="ant-form-item-required"
          htmlFor="cvc"
          onClick={[Function]}
          title="CVC"
        >
          CVC
        </label>
      </div>
      <div
        className="ant-col ant-form-item-control-wrapper"
      >
        <div
          className="ant-form-item-control"
        >
          <span
            className="ant-form-item-children"
          >
            <input
              className="ant-input"
              data-__field={
                Object {
                  "name": "cvc",
                }
              }
              data-__meta={
                Object {
                  "getValueFromEvent": [Function],
                  "name": "cvc",
                  "originalProps": Object {
                    "disabled": false,
                    "maxLength": 3,
                    "placeholder": "123",
                    "type": "text",
                  },
                  "ref": null,
                  "rules": Array [
                    Object {
                      "message": "Required",
                      "required": true,
                    },
                    Object {
                      "len": 3,
                      "message": "Too short",
                    },
                  ],
                  "trigger": "onChange",
                  "validate": Array [
                    Object {
                      "rules": Array [
                        Object {
                          "message": "Required",
                          "required": true,
                        },
                        Object {
                          "len": 3,
                          "message": "Too short",
                        },
                      ],
                      "trigger": Array [
                        "onSubmit",
                      ],
                    },
                  ],
                  "validateTrigger": "onSubmit",
                  "valuePropName": "value",
                }
              }
              disabled={false}
              id="cvc"
              maxLength={3}
              onChange={[Function]}
              onKeyDown={[Function]}
              onSubmit={[Function]}
              placeholder="123"
              type="text"
              value=""
            />
          </span>
        </div>
      </div>
    </div>
    <div
      className="ant-row ant-form-item creditCardForm-expiryFormItem"
    >
      <div
        className="ant-col ant-form-item-label"
      >
        <label
          className="ant-form-item-required"
          htmlFor="expiry"
          onClick={[Function]}
          title="Expiry"
        >
          Expiry
        </label>
      </div>
      <div
        className="ant-col ant-form-item-control-wrapper"
      >
        <div
          className="ant-form-item-control"
        >
          <span
            className="ant-form-item-children"
          >
            <input
              className="ant-input"
              data-__field={
                Object {
                  "name": "expiry",
                }
              }
              data-__meta={
                Object {
                  "getValueFromEvent": [Function],
                  "name": "expiry",
                  "originalProps": Object {
                    "disabled": false,
                    "maxLength": 5,
                    "placeholder": "12/34",
                    "type": "text",
                  },
                  "ref": null,
                  "rules": Array [
                    Object {
                      "message": "Required",
                      "required": true,
                    },
                    Object {
                      "len": 5,
                      "message": "Too short",
                    },
                  ],
                  "trigger": "onChange",
                  "validate": Array [
                    Object {
                      "rules": Array [
                        Object {
                          "message": "Required",
                          "required": true,
                        },
                        Object {
                          "len": 5,
                          "message": "Too short",
                        },
                      ],
                      "trigger": Array [
                        "onSubmit",
                      ],
                    },
                  ],
                  "validateTrigger": "onSubmit",
                  "valuePropName": "value",
                }
              }
              disabled={false}
              id="expiry"
              maxLength={5}
              onChange={[Function]}
              onKeyDown={[Function]}
              onSubmit={[Function]}
              placeholder="12/34"
              type="text"
              value=""
            />
          </span>
        </div>
      </div>
    </div>
  </div>
  <div
    className="ant-row ant-form-item"
  >
    <div
      className="ant-col ant-form-item-control-wrapper"
    >
      <div
        className="ant-form-item-control"
      >
        <span
          className="ant-form-item-children"
        >
          <button
            className="ant-btn creditCardForm-submitButton"
            onClick={[Function]}
            type="button"
          >
            <span>
              Submit
            </span>
          </button>
        </span>
      </div>
    </div>
  </div>
</form>
`;
